sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/client/DefaultHttpClient.java

```diff
@@ -197,37 +205,100 @@ private void sendBody(HttpRequest httpRequest, HttpURLConnection connection) {
      * @return A HttpResponse object
      */
     private Response<?> receiveResponse(HttpRequest httpRequest, HttpURLConnection connection) {
-        try {
-            int responseCode = connection.getResponseCode();
-            HttpHeaders responseHeaders = getResponseHeaders(connection);
+        HttpHeaders responseHeaders = getResponseHeaders(connection);
+        HttpResponse<?> httpResponse = createHttpResponse(httpRequest, connection);
 
+        if (isTextEventStream(responseHeaders)) {
             ServerSentEventListener listener = httpRequest.getServerSentEventListener();
 
-            if (listener == null && isTextEventStream(responseHeaders)) {
+            if (listener == null) {
                 throw LOGGER.logThrowableAsError(new RuntimeException(NO_LISTENER_ERROR_MESSAGE));
             }
 
-            if (connection.getErrorStream() == null && isTextEventStream(responseHeaders)) {
-                processTextEventStream(httpRequest, httpRequestConsumer -> this.send(httpRequest), connection.getInputStream(), listener, LOGGER);
-                return new HttpResponse<>(httpRequest, responseCode, responseHeaders, null);
+            if (connection.getErrorStream() == null) {
+                try {
+                    processTextEventStream(httpRequest, httpRequestConsumer ->
+                        this.send(httpRequest), connection.getInputStream(), listener, LOGGER);
+                } catch (IOException e) {
+                    throw LOGGER.logThrowableAsError(new UncheckedIOException(e));
+                } finally {
+                    connection.disconnect();
+                }
             }
+        } else {
+            ResponseBodyMode responseBodyMode = httpRequest.getMetadata().getResponseBodyMode();
 
-            AccessibleByteArrayOutputStream outputStream = getAccessibleByteArrayOutputStream(connection);
-            return new HttpResponse<>(httpRequest, responseCode, responseHeaders, BinaryData.fromByteBuffer(outputStream.toByteBuffer()));
+            if (responseBodyMode == null) {
+                HttpHeader contentType = httpResponse.getHeaders().get(CONTENT_TYPE);
+                String contentTypeString = contentType == null ? null : contentType.getValue();
+
+                if (APPLICATION_OCTET_STREAM.equalsIgnoreCase(contentTypeString)) {
+                    responseBodyMode = STREAM;
+                } else {
+                    responseBodyMode = BUFFER;
+                }
+
+                httpRequest.getMetadata().setResponseBodyMode(responseBodyMode);
+            }
+
+            switch (responseBodyMode) {
+                case IGNORE:
+                    HttpResponseAccessHelper.setBody(httpResponse, EMPTY_BODY);
+
+                    connection.disconnect();
+
+                    break;
+                case STREAM:
+                    streamResponseBody(httpResponse, connection);
+
+                    break;
+                case BUFFER:
+                case DESERIALIZE:
```

Comment:
Are we not eagerly deserializing when this mode is set? If not, do we need this mode?

Reply:
We deserialize the contents at a later point in `HttpResponseBodyDecoder.handleRestResponseReturnType()`.

Reply:
The only case where we don't close the connection is for `STREAM`.